### PR TITLE
feat: improve embedding of swapbox

### LINKED ISSUE #334: Embed: add url parameters
URL parameters (all optional, we default):

- `pairID`: (default: `BTC/BTC`)
- `type`: (default: `reversesubmarine`)
- `bip21`: Overwrites parameters above the same way web app swap box does. We only need address & optional amount. Let's not forget `liquidnetwork:`
- `invoice`: overwrites parameters above. (no default)
- `referral`: partner referral id (default: boltz_webapp_embed)
- `css`: style sheet for swap box customization

_Originally posted by @kilrau in https://github.com/BoltzExchange/boltz-web-app/issues/328#issuecomment-1820505661_

### PR DESCRIPTION
**Pull request by dni**
closes #334

adds invoice, address, asset, sendAmount, receiveAmount

### CODE REVIEW

**Comment by kilrau** [comment]
Shall we push this one over the finish line? We recently had an integration that wanted "a simplified webview with prefilled swap data"

**Comment by dni** [comment]
created an example embed page http://localhost:5174/boltz-webapp-embed/

**Comment by tohrxyz** [comment]
really looking forward to this merging to main, because it's one of the fundamental parts that our project needs right now! :)

**Review by kilrau — CHANGES_REQUESTED** [review]
- `setInvoice` / `setAddress` -> `setDestination`
- `assetSend` / `assetReceive` -> `sendAsset` / `receiveAsset` to bring naming in line

File: src/context/Global.tsx [hunk 1]
```diff
@@ -275,19 +275,15 @@ const GlobalProvider = (props: { children: any }) => {
     detectWebLNProvider().then((state: boolean) => setWebln(state));
     setWasmSupported(checkWasmSupported());
 
-    // check referral
-    const refParam = new URLSearchParams(window.location.search).get("ref");
+    // check referal
```
Comment (michael1011) [inline src/context/Global.tsx:278]:
```suggestion
    // check referral
```

File: src/utils/urlparams.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+export const detectEmbedParam = (): boolean => {
+    const param = searchParams.get("embed");
+    return param && param === "1";
+};
+
+export const detectUrlParam = (param: string): string => {
+    const searchParam = searchParams.get(param);
+    if (searchParam) {
+        resetUrlParam(param);
```
Comment (michael1011) [inline src/utils/urlparams.ts:11]: Why are we removing it?

File: src/context/Global.tsx [hunk 3]
```diff
@@ -275,19 +275,15 @@ const GlobalProvider = (props: { children: any }) => {
...
-            {},
-            document.title,
-            window.location.pathname,
-        );
     }
 
-    if (detectEmbedded()) {
+    if (detectEmbedParam()) {
         setEmbedded(true);
+        setHideHero(true);
```
Comment (michael1011) [inline src/context/Global.tsx:286]: Shouldn't that be an effect of `setEmbedded(true)`?

File: src/context/Create.tsx [hunk 4]
```diff
@@ -132,6 +133,42 @@ const CreateProvider = (props: { children: any }) => {
     const [boltzFee, setBoltzFee] = createSignal(0);
     const [minerFee, setMinerFee] = createSignal(0);
 
+    // embed
```
Comment (michael1011) [inline src/context/Create.tsx:136]: You don't have to embed to use those

File: src/context/Create.tsx [hunk 5]
```diff
@@ -132,6 +133,42 @@ const CreateProvider = (props: { children: any }) => {
...
+    const assetSendParam = detectUrlParam("assetSend");
+    const assetReceiveParam = detectUrlParam("assetReceive");
+    if (
+        assetSendParam &&
+        assetSendParam !== "" &&
+        assetReceiveParam &&
+        assetReceiveParam !== "" &&
+        [BTC, LN, LBTC, RBTC].includes(assetSendParam) &&
+        [BTC, LN, LBTC, RBTC].includes(assetReceiveParam) &&
+        assetSendParam !== assetReceiveParam
```
Comment (michael1011) [inline src/context/Create.tsx:155]: Why not just check if that pair is valid based on the API responses we fetch on load?

File: src/context/Create.tsx [hunk 6]
```diff
@@ -132,6 +133,42 @@ const CreateProvider = (props: { children: any }) => {
...
+    ) {
+        setAssetSend(assetSendParam);
+        setAssetReceive(assetReceiveParam);
+    }
+    const sendAmountParam = detectUrlParam("sendAmount");
+    const isSendAmountValid = sendAmountParam && sendAmountParam !== "";
+    if (isSendAmountValid) {
+        setSendAmount(BigNumber(sendAmountParam));
+    }
+    // only set receive amount if send amount is set
```
Comment (michael1011) [inline src/context/Create.tsx:165]:
```suggestion
    // only set receive amount if send amount is not set
```

**Review by michael1011 — CHANGES_REQUESTED** [review]
All URL query params we want people to use need to be documented

**Comment by kilrau** [comment]
Let's go @dni , @tohrxyz and team are waiting for this

**Comment by tohrxyz** [comment]
I'm looking forward to have a few examples ready of URLs with different strategies for swaps. This will help also future devs with integrating with Boltz.

So far good job and I'm super happy with the responsiveness of you guys! :)

**Review by maybeast — APPROVED** [review]

**Pull request merged**